fs.go

```diff
@@ -17,96 +17,521 @@
+					return volumeDir, nil
+				}
+			}
+		}
+		return volumeDir, errVolumeNotFound
+	} else if os.IsPermission(err) {
+		return volumeDir, errVolumeAccessDenied
 	}
-	return bucket, nil
+	return volumeDir, err
 }
 
-func checkDiskFree(diskPath string, minFreeDisk int64) error {
-	di, e := disk.GetInfo(diskPath)
-	if e != nil {
+// Make a volume entry.
+func (s fsStorage) MakeVol(volume string) (err error) {
+	volumeDir, err := s.checkVolumeArg(volume)
+	if err == nil {
+		// Volume already exists, return error.
+		return errVolumeExists
+	}
+
+	// Validate if disk is free.
+	if e := checkDiskFree(s.diskPath, s.minFreeDisk); e != nil {
 		return e
 	}
-	// Remove 5% from total space for cumulative disk space used for journalling, inodes etc.
-	availableDiskSpace := (float64(di.Free) / (float64(di.Total) - (0.05 * float64(di.Total)))) * 100
-	if int64(availableDiskSpace) <= minFreeDisk {
-		return RootPathFull{Path: diskPath}
+
+	// If volume not found create it.
+	if err == errVolumeNotFound {
+		// Make a volume entry.
+		return os.Mkdir(volumeDir, 0700)
+	}
+
+	// For all other errors return here.
+	return err
+}
+
+// removeDuplicateVols - remove duplicate volumes.
+func removeDuplicateVols(vols []VolInfo) []VolInfo {
+	length := len(vols) - 1
+	for i := 0; i < length; i++ {
+		for j := i + 1; j <= length; j++ {
+			if vols[i].Name == vols[j].Name {
+				// Pick the latest volume, if there is a duplicate.
+				if vols[i].Created.Sub(vols[j].Created) > 0 {
+					vols[i] = vols[length]
+				} else {
+					vols[j] = vols[length]
+				}
+				vols = vols[0:length]
+				length--
+				j--
+			}
+		}
+	}
+	return vols
+}
+
+// ListVols - list volumes.
+func (s fsStorage) ListVols() (volsInfo []VolInfo, err error) {
+	files, err := ioutil.ReadDir(s.diskPath)
+	if err != nil {
+		return nil, err
+	}
+	for _, file := range files {
+		if !file.IsDir() {
+			// If not directory, ignore all file types.
+			continue
+		}
+		// Volname on case sensitive fs backends can come in as
+		// capitalized, but object layer cannot consume it
+		// directly. Convert it as we see fit.
+		volName := strings.ToLower(file.Name())
+		// Modtime is used as created time.
+		createdTime := file.ModTime()
+		volInfo := VolInfo{
+			Name:    volName,
+			Created: createdTime,
+		}
+		volsInfo = append(volsInfo, volInfo)
+	}
+	// Remove duplicated volume entries.
+	volsInfo = removeDuplicateVols(volsInfo)
+	return volsInfo, nil
+}
+
+// StatVol - get volume info.
+func (s fsStorage) StatVol(volume string) (volInfo VolInfo, err error) {
+	// Verify if volume is valid and it exists.
+	volumeDir, err := s.checkVolumeArg(volume)
+	if err != nil {
+		return VolInfo{}, err
+	}
+	// Stat a volume entry.
+	var st os.FileInfo
+	st, err = os.Stat(volumeDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return VolInfo{}, errVolumeNotFound
+		}
+		return VolInfo{}, err
+	}
+	// Modtime is used as created time since operating systems lack a
```

Comment:
As a note, its not operating system limitation. http://serverfault.com/questions/351836/whats-birth-field-in-stat-command-output.  golang provides ModTime() only.